Integrate web terminal to user cluster

KhizerRehan opened the pull request:
### What this PR does / why we need it

This feature allows to have terminal access for a user cluster. Terminal can be access via WebTerminal button on Cluster Details page.

- Terminal can be opened as within same cluster details page 
- Terminal can be open as "Open in new tab" from Terminal 

<img width="1435" alt="terminal-button" src="https://user-images.githubusercontent.com/17727069/169294810-472fbab4-9b54-436d-ac78-37b2848a84b7.png">

**Other functionality:**

- Allows to reconnect to terminal in case terminal connection is closed:

**Light Mode**

- Cluster Details Page:

**Initial WS connection**

https://user-images.githubusercontent.com/17727069/194856255-edcc71e9-0952-426c-b085-f7211237385b.mp4


**Authentication Succeeded and Connected to Shell**

https://user-images.githubusercontent.com/17727069/194856227-e7605e86-fbf5-484a-b793-fe3eea4cadac.mp4



- As page

![full-apge](https://user-images.githubusercontent.com/17727069/194856113-d945b634-d72b-4b38-807c-81b019b0c9be.png)



### Which issue(s) this PR fixes
<!--
Use one of the GitHub's keywords to link connected Issues/PRs.
Keyword list: https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword
-->

Closes https://github.com/kubermatic/dashboard/issues/4432

### Special notes for your reviewer
<!-- Remove if not needed -->

Button is only visible when "Enable OIDC Kubeconfig" **feature gate** is enabled.

### Release note
<!--
Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just leave "NONE".
-->
```release-note
Add Web terminal support for user clusters
```

Waseem826 commented on src/app/shared/components/xterm-terminal/terminal-toolbar/component.ts:
```diff
@@ -0,0 +1,40 @@
+  @Input() showCloseIcon = true;
+  @Input() showOpenInNewTabIcon = true;
+
+  @Output() closeTerminal = new EventEmitter<boolean>();
```
Same as previous comment, `void` instead of `boolean`

KhizerRehan replied:
The only reason for using `type` was to make it more clearer when reading code `void` will do the work since we know we have clicked the button and event is supposed to be done there are no `If and else` checks so event 
`.next`.
okay i can fix that.

Waseem826 commented on src/app/shared/components/xterm-terminal/terminal-toolbar/component.ts:
```diff
@@ -0,0 +1,40 @@
+
+  constructor() {}
+
+  ngOnInit(): void {}
```
Remove empty `ngOnInit` and `constructor` here as well

KhizerRehan replied:
IMHO, constructor is part of class and it is default generated that must be there but yes empty lifecycle hook should be removed. thoughts?

KhizerRehan replied:
i guess code coverage is also not covering which causes pipeline for coverage to fail also

KhizerRehan replied:
fixed

Waseem826 commented on src/app/style.scss:
```diff
@@ -28,6 +28,12 @@ mat-sidenav-container {
+  // Note:
+  // WebTerminal is hidden under z-index mat-sidenav
+  & mat-sidenav.mat-drawer.mat-drawer-side {
+    z-index: 1;
```
This style does not have any affect

KhizerRehan replied:
correct! This is because previously, the terminal i was implementing as `FIXED` at bottom and placing in root level that was not possible due to some UI limitation/structure + Terminal was more specific so In order to show as overlay `z-index` was causing issue to hide overlay underneath.

But now it is part of cluster module and only visible content area.

Good catch! i will remove it.

KhizerRehan replied:
fixed

KhizerRehan commented on src/app/shared/components/xterm-terminal/terminal-status-bar/template.html:
```diff
@@ -0,0 +1,59 @@
+            (click)="onCloseTerminal()">
+      <span>Close</span>
+    </button>
+    <button mat-raised-button
```
- I added mat-raised-button -> "Rectangular contained button w/ elevation" i wanted to have slight elevation to look more like it clickable but yes i can use `mat-flat-button`

KhizerRehan replied:
fixed

KhizerRehan commented on src/app/shared/components/xterm-terminal/terminal-status-bar/component.ts:
```diff
@@ -0,0 +1,39 @@
+
+  constructor() {}
+
+  ngOnInit(): void {}
```
https://github.com/kubermatic/dashboard/pull/4492#discussion_r882362504

KhizerRehan commented on angular.json:
```diff
@@ -32,6 +32,7 @@
               ]
             },
             "styles": [
+              "node_modules/xterm/css/xterm.css",
```
fixed!

KhizerRehan commented on src/app/cluster/details/cluster/component.ts:
```diff
@@ -128,7 +129,9 @@ export class ClusterDetailsComponent implements OnInit, OnDestroy {
     readonly settings: SettingsService
-  ) {}
+  ) {
+    this.showTerminal = false;
```
fixed

KhizerRehan commented on src/app/shared/components/xterm-terminal/terminal-status-bar/style.scss:
```diff
@@ -0,0 +1,50 @@
+
+  // Close button
+  & .close-terminal-button {
+    background-color: #787b7e4d;
```
fixed

KhizerRehan commented on src/app/shared/components/xterm-terminal/terminal-status-bar/style.scss:
```diff
@@ -0,0 +1,44 @@
+  & button[mat-raised-button], {
+    @include mixins.size(7.812rem, 2rem);
+
+    color: white;
```
already fixed

KhizerRehan commented on src/app/shared/components/xterm-terminal/component.ts:
```diff
@@ -0,0 +1,178 @@
+    });
+
+    // Subscription connection success
+    this.websocketService.getWebSocketOnConnectObservable$().subscribe((isConnected: boolean) => {
```
fixed

KhizerRehan commented on src/app/cluster/details/cluster/component.ts:
```diff
@@ -527,6 +530,25 @@ export class ClusterDetailsComponent implements OnInit, OnDestroy {
+  }
+
+  onCloseTerminal(isClosed: boolean) {
+    if (isClosed) {
```
Added for type safety reason. But it will always be false and secondly we are sure that user has pressed the close button either from toolbar/status bar so it will be `true` means -> user has performed action to close terminal which is `true` 
I have got your point it is fixed: [`2476044` (#4492)](https://github.com/kubermatic/dashboard/pull/4492/commits/2476044226467a03f740c2a7a0800d1a93b71c76)

KhizerRehan commented on src/app/shared/components/xterm-terminal/component.ts:
```diff
@@ -0,0 +1,206 @@
+    if (reconnect) {
+      // Re-initialize new terminal
+      this.initTerminal();
+      this.websocketService.connect(this.getConnectionString());
```
fixed [`64082fd` (#4492)](https://github.com/kubermatic/dashboard/pull/4492/commits/64082fd54fee325820caaf859c4c8b25b15435da)

KhizerRehan commented on src/app/shared/components/xterm-terminal/component.ts:
```diff
@@ -0,0 +1,206 @@
+    delayFn();
+
+    // Event listeners binding
+    this._resizeEventlistenerFn = this._renderer.listen(window, 'resize', () => {
```
fixed [`2ded947` (#4492)](https://github.com/kubermatic/dashboard/pull/4492/commits/2ded94725d4f6b198c466b6af48897cd985dbc0c)

KhizerRehan commented on src/app/core/services/websocket.ts:
```diff
@@ -0,0 +1,101 @@
+export class WebsocketService {
+  private readonly _wsRoot = environment.wsRoot;
+  private _socket$: WebSocketSubject<any>;
+  private _onConnect: Subject<boolean> = new Subject<boolean>();
```
Fixed: [`2e1c5cf` (#4492)](https://github.com/kubermatic/dashboard/pull/4492/commits/2e1c5cf475472d38d6f25859a0d1bbab83db55ee)

KhizerRehan commented on src/app/cluster/details/cluster/web-terminal/component.ts:
```diff
@@ -0,0 +1,25 @@
+  encapsulation: ViewEncapsulation.None,
+})
+export class WebTerminalComponent {
+  constructor() {}
```
fixed

KhizerRehan commented on src/app/cluster/details/cluster/component.ts:
```diff
@@ -527,6 +530,25 @@ export class ClusterDetailsComponent implements OnInit, OnDestroy {
+  }
+
+  onOpenInNewTab(isOpenInNewTab: boolean) {
+    if (isOpenInNewTab) {
```
Fixed

maciaszczykm commented on src/app/cluster/details/cluster/overlay-terminal/template.html:
```diff
@@ -0,0 +1,28 @@
+<div id="overlay-terminal">
+  <!---------------------------->
+  <!-- Terminal Toolbar -->
+  <!---------------------------->
```
It doesn't bring any explanation, please remove.

maciaszczykm commented on src/app/cluster/details/cluster/overlay-terminal/template.html:
```diff
@@ -0,0 +1,28 @@
+
+  <!---------------------------->
+  <!-- Terminal -->
+  <!---------------------------->
```
As above.

maciaszczykm commented on src/app/shared/components/xterm-terminal/component.ts:
```diff
@@ -0,0 +1,215 @@
+    this.terminal.write('\x1b[37mWEB TERMINAL\x1b\r\n\n');
+    this.terminal.write('\x1b[37mWelcome to Cloud Shell! Type "help" to get started.\r\n\n');
+    this.terminal.write(
+      `\x1b[37mYour Cloud Platform project in this session is set to \x1b[1;34m${clusterName}\x1B[0m\n\n\r`
```
@cschieder Should we write this every time? I think it should be really short if we want to add something.

maciaszczykm commented on src/app/shared/components/xterm-terminal/component.ts:
```diff
@@ -0,0 +1,215 @@
+    this.terminalDisposed.emit();
+  }
+
+  onReconnect() {
```
Return type is missing here and in many other methods.

maciaszczykm replied:
@KhizerRehan this is not resolved, check all your changes.

KhizerRehan replied:
Fixed commit [`6e41851` (#4492)](https://github.com/kubermatic/dashboard/pull/4492/commits/6e41851d63882950dd1b1a529f0d08a56556364c)

maciaszczykm replied:
I've added comments in places where it is still not done, i.e. https://github.com/kubermatic/dashboard/pull/4492/files#diff-cd8604576285d61d53d3648ae03e91dbe49e58aa6f262733dccadfbc79316dd7R531.

maciaszczykm commented on src/app/shared/components/xterm-terminal/template.html:
```diff
@@ -0,0 +1,30 @@
+<!---------------------------->
+<!--Status Toolbar-->
+<!---------------------------->
+<ng-container *ngIf="isConnectionLost">
```
Why using additional ng-container?

maciaszczykm commented on src/app/shared/model/Terminal.ts:
```diff
@@ -0,0 +1,23 @@
+  Data?: string;
+  SessionID?: string;
+  Rows?: number;
+  Cols?: number;
```
Is it coming from some API or is it internal? If internal please follow our naming conventions.

KhizerRehan replied:
This API format
https://github.com/kubermatic/kubermatic/blob/master/pkg/handler/websocket/terminal.go#L63

maciaszczykm commented on src/app/shared/components/xterm-terminal/terminal-status-bar/style.scss:
```diff
@@ -0,0 +1,43 @@
+    margin: 0;
+  }
+
+  // Close button
```
The class name is self-explanatory, comments like this are not required.

maciaszczykm commented on src/app/shared/components/xterm-terminal/style.scss:
```diff
@@ -0,0 +1,41 @@
+
+  ::-webkit-scrollbar-corner,
+  ::-webkit-scrollbar-track {
+    background: none !important;
```
Avoid using `!important`.

maciaszczykm commented on src/app/shared/components/xterm-terminal/component.ts:
```diff
@@ -0,0 +1,215 @@
+        foreground: '#f8f8f8',
+        background: '#2b3035',
+      },
+    });
```
Export config to readonly variable.

KhizerRehan replied:
const are created at top of the file to have all Terminal related configurations.

maciaszczykm commented:
Comments based on the screenshots and video:
- Should the terminal icon be so big? It looks strange to me.
- There is no padding inside the terminal.
- Generally the UI looks different than [on the designs provided](https://www.figma.com/file/dJQKgUPzIDND36o2Nq8dct/Loodse-UI?node-id=6670%3A54116) by @cschieder. That needs to be resolved before we can approve that.

<img width="724" alt="Zrzut ekranu 2022-05-26 o 12 53 26" src="https://user-images.githubusercontent.com/2823399/170474399-2f79194c-dbac-4e37-8314-4516aa8f06f4.png">

maciaszczykm reviewed (changes requested):
As mentioned above.

KhizerRehan commented:
> Comments based on the screenshots and video:
> 
> * Should the terminal icon be so big? It looks strange to me.
> * There is no padding inside the terminal.
> * Generally the UI looks different than [on the designs provided](https://www.figma.com/file/dJQKgUPzIDND36o2Nq8dct/Loodse-UI?node-id=6670%3A54116) by @cschieder. That needs to be resolved before we can approve that.
> 
> <img alt="Zrzut ekranu 2022-05-26 o 12 53 26" width="724" src="https://user-images.githubusercontent.com/2823399/170474399-2f79194c-dbac-4e37-8314-4516aa8f06f4.png">

<img width="458" alt="name@cloudshell" src="https://user-images.githubusercontent.com/17727069/170491570-feb57142-b3ee-4035-8db3-0a83f093bb89.png">

with "bash-5.1$ " appended on each returned output? This is something IMHO, can be handled at backend side

@zreigz @maciaszczykm

KhizerRehan commented:
Updated Screenshot of terminal

<img width="1213" alt="updated screenshot" src="https://user-images.githubusercontent.com/17727069/170650646-7545c5fd-0257-4a6e-b87a-7833ce810ec2.png">

cc: @cschieder

maciaszczykm commented:
> with "bash-5.1$ " appended on each returned output? This is something IMHO, can be handled at backend side

Yes, this would need to be adjusted in the Docker image that we use for providing terminal.

cc @zreigz

maciaszczykm commented:
@KhizerRehan `WEB TERMINAL` inside terminal is redundant, please remove it.

maciaszczykm commented on src/app/shared/components/xterm-terminal/terminal-status-bar/style.scss:
```diff
@@ -0,0 +1,40 @@
+  }
+
+  .km-icon-warning {
+    @include mixins.size(20px, 20px, true);
```
`@include mixins.size(20px, $force: true);`

maciaszczykm commented on src/app/shared/components/xterm-terminal/terminal-status-bar/theme.scss:
```diff
@@ -0,0 +1,32 @@
+
+    // 'X' icon button
+    & .close-terminal-button:hover {
+      background-color: map.get($colors, terminal-status-bar-close-button-hover) !important;
```
In most of the cases `!important` can be replaced with usage of specific class selectors. Also, the comment above the rule is unnecessary.

maciaszczykm commented on src/app/shared/components/xterm-terminal/terminal-status-bar/theme.scss:
```diff
@@ -0,0 +1,32 @@
+    }
+
+    & .button-border-box {
+      border: 0.7px solid map.get($colors, terminal-status-bar-button-border-color);
```
Why not 1px?

maciaszczykm commented on src/app/shared/components/xterm-terminal/terminal-toolbar/style.scss:
```diff
@@ -0,0 +1,44 @@
+  background-color: #2b3035;
+  box-shadow: 2px 4px 29px -6px black;
+  box-sizing: border-box;
+  color: #fff;
```
Avoid any colors inside style files and put them in theme files.

maciaszczykm commented on src/app/shared/components/xterm-terminal/terminal-toolbar/style.scss:
```diff
@@ -0,0 +1,44 @@
+.toolbar-container {
+  @include mixins.size(100%, 4rem);
+
+  align-items: center;
```
It would be nice to use flex-layout in the template for this.

KhizerRehan replied:
removed

maciaszczykm commented on src/app/shared/components/xterm-terminal/terminal-toolbar/style.scss:
```diff
@@ -0,0 +1,44 @@
+  }
+
+  .km-icon-terminal {
+    @include mixins.size(20px, 20px, true);
```
`@include mixins.size(20px, $force: true);`

maciaszczykm commented on src/app/cluster/details/cluster/component.ts:
```diff
@@ -527,6 +528,21 @@ export class ClusterDetailsComponent implements OnInit, OnDestroy {
     return !!this.cluster?.spec?.cniPlugin && this.cluster?.spec?.cniPlugin?.type !== CNIPlugin.None;
   }
 
+  toggleTerminal() {
```
Missing method signature.

maciaszczykm commented on src/app/cluster/details/cluster/component.ts:
```diff
@@ -527,6 +528,21 @@ export class ClusterDetailsComponent implements OnInit, OnDestroy {
+    this.showTerminal = !this.showTerminal;
+  }
+
+  onCloseTerminal() {
```
As above.

maciaszczykm commented on src/app/cluster/details/cluster/component.ts:
```diff
@@ -527,6 +528,21 @@ export class ClusterDetailsComponent implements OnInit, OnDestroy {
+    this.toggleTerminal();
+  }
+
+  onOpenInNewTab() {
```
As above.

maciaszczykm commented on src/app/core/services/websocket.ts:
```diff
@@ -0,0 +1,97 @@
+    })
+  );
+
+  get isSocketClosed() {
```
As above.

KhizerRehan commented:
/retest

ahmedwaleedmalik commented:
/test pre-dashboard-test-e2e-ce

maciaszczykm commented:
/hold

@Waseem826 You have already approved it. Adding hold so I can do another round of review tomorrow before approving from my side.

@KhizerRehan Please avoid resolving my comments so I can do it after checking that everything was done.

maciaszczykm commented:
/hold

@Waseem826 You have already approved it. Adding hold so I can do another round of review tomorrow before approving from my side.

maciaszczykm commented on src/app/cluster/details/cluster/overlay-terminal/style.scss:
```diff
@@ -0,0 +1,21 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#overlay-terminal {
```
Use classes to style elements.

maciaszczykm commented:
<img width="969" alt="Zrzut ekranu 2022-05-30 o 11 18 24" src="https://user-images.githubusercontent.com/2823399/170960635-2aa723f4-7b96-44ea-89f3-53f575976da3.png">

@cschieder A little thing related to the design - all other buttons contain verbs, should this be i.e. "Open Web Terminal"?

maciaszczykm commented:
<img width="91" alt="Zrzut ekranu 2022-05-30 o 11 23 24" src="https://user-images.githubusercontent.com/2823399/170961287-ba722757-1da7-458f-9cf6-697fd1fdbae1.png">

Part of the shadow is under the menu.

maciaszczykm commented:
<img width="249" alt="Zrzut ekranu 2022-05-30 o 11 25 51" src="https://user-images.githubusercontent.com/2823399/170961784-c09eb2e6-bc0b-45b0-a468-73504dd9b708.png">

Make it look like in the design:

<img width="164" alt="Zrzut ekranu 2022-05-30 o 11 26 21" src="https://user-images.githubusercontent.com/2823399/170961927-4cc15d91-c221-433c-9c5e-d44a124bf2a3.png">

maciaszczykm commented:
If that's possible hide the blinking cursor when the connection was lost as it is not possible to write anyways.

maciaszczykm commented:
<img width="575" alt="Zrzut ekranu 2022-05-30 o 11 35 27" src="https://user-images.githubusercontent.com/2823399/170963675-9c8a76dc-94f3-4c57-99c4-6e33a0d06679.png">

The info about cluster is missing, there is no back button too.

maciaszczykm commented:
<img width="2032" alt="Zrzut ekranu 2022-05-30 o 11 37 42" src="https://user-images.githubusercontent.com/2823399/170964071-a156ee3b-e4be-4662-be90-fac51e7fd8f0.png">

When opening in a new tab terminal is too big. It should fit the browser window with footer. I saw that you have used 100vh in the code, so that is probably it.

maciaszczykm commented on src/app/shared/components/xterm-terminal/terminal-toolbar/style.scss:
```diff
@@ -0,0 +1,35 @@
+    font-size: 1rem;
+    line-height: 1.5rem;
+    margin: 5px 0;
+    text-transform: uppercase;
```
In the design it's not uppercase.

maciaszczykm commented on src/app/cluster/details/cluster/overlay-terminal/template.html:
```diff
@@ -0,0 +1,22 @@
+-->
+
+<div id="overlay-terminal">
+  <km-terminal-toolbar (openInNewTab)="onOpenInNewTab()"
```
Why it is not put inside km-terminal component?

maciaszczykm commented on package-lock.json:
```diff
@@ -1,7 +1,7 @@
   "name": "kubermatic-dashboard",
   "version": "2.20.0",
-  "lockfileVersion": 2,
+  "lockfileVersion": 1,
```
Update your npm to keep new version of lock file.

floreks commented on src/app/shared/components/xterm-terminal/component.ts:
```diff
@@ -0,0 +1,227 @@
+  selector: 'km-terminal',
+  templateUrl: './template.html',
+  styleUrls: ['./style.scss'],
+  encapsulation: ViewEncapsulation.None,
```
Do not use it.

floreks reviewed (changes requested):
Done the first batch of review. Will do another one tomorrow.

cschieder commented:
1. Is it by chance possible to cut off the shadow at the sides? ![image](https://user-images.githubusercontent.com/49145827/170972415-c478d248-40af-4cb5-bca5-a9d897831c72.png)


![Terminal_2](https://user-images.githubusercontent.com/49145827/170972670-0581def6-1c30-463d-a954-b39bd1dfb73e.png)


2. Please remove the shadow for the stand alone terminal and add the typical card drop shadow and border radius.

KhizerRehan commented:
/retest

KhizerRehan commented on src/app/cluster/details/cluster/web-terminal/component.ts:
```diff
@@ -0,0 +1,23 @@
+  styleUrls: ['style.scss'],
+  encapsulation: ViewEncapsulation.None,
+})
+export class WebTerminalComponent {}
```
Basically implementation is such that we have a generic component called:
XTermTerminalComponent which is more sort of reusable component

We have two wrapper component whos stylng differes
- As Overlay in ClusterDeatilsComponent
- As Page when open as new tab

- Styling differs 
- Terminal size differers

therefore in order to sep both instead of internally adding many changes based on condtions i decided to have sep wrapper one will act as overlay and other as page

KhizerRehan commented on src/app/shared/components/xterm-terminal/style.scss:
```diff
@@ -0,0 +1,38 @@
+// By wrapping styles with "selector" will prevent styles to go up the chain to parent and sibling components.
+// It will only effect styles specified here will go down to children components only.
+
+km-terminal {
```
According to my understandng
`:host` works when `ViewEncapuslation.shadowDOM` which is default but i need to target to xterm module css which was not effecting due to shadowDOM i found a useful [stackoverflow (https://stackoverflow.com/a/53843161/6618218)

Also added Note: [L#](https://github.com/kubermatic/dashboard/pull/4492/files/5faf26431329d90d663e2929a95b575d1fe0c791#diff-fb158a04cc4b5e25ae2ea0688a4c226d47fa8f4a264dab158a5a34db538c26cfR15)

KhizerRehan commented on src/app/cluster/details/cluster/web-terminal/component.ts:
```diff
@@ -0,0 +1,23 @@
+  selector: 'km-web-terminal',
+  templateUrl: './template.html',
+  styleUrls: ['style.scss'],
+  encapsulation: ViewEncapsulation.None,
```
[As discussed below: ](https://github.com/kubermatic/dashboard/pull/4492/files/5faf26431329d90d663e2929a95b575d1fe0c791#r891976155)

KhizerRehan commented:
/PTAL 

Note: Instead of making individual feedback as "Mark as Resolved" i have fixed it and which are not i have added comment on that. Let me know if you found still some issues :)

Talha-Jamil-TJ commented on src/app/cluster/details/cluster/web-terminal/component.ts:
```diff
@@ -0,0 +1,86 @@
+      });
+  }
+
+  isInitialized(): boolean {
```
can we move this type of login in a getter?

KhizerRehan replied:
I just kept consistent with other features! 

e.g `MachineDeploymentDetailsPage`

Talha-Jamil-TJ commented on src/app/cluster/details/cluster/template.html:
```diff
@@ -446,3 +455,8 @@
+
+
+<km-overlay-terminal *ngIf="showTerminal"
+                     (close)="toggleTerminal()">
```
please add id to this element, so it would be easy for testing.

KhizerRehan replied:
There is no unit testing done! so therefore it is not added.

maciaszczykm replied:
I agree with @KhizerRehan here. Let's not add things that we don't use. If the test will be written in the future it is really simple to add new ID. Otherwise, we will end up with a lot of redundant things in the code.

Talha-Jamil-TJ commented on src/app/shared/components/xterm-terminal/component.ts:
```diff
@@ -0,0 +1,252 @@
+    });
+
+    // XTerminal Properties
+    this.terminal.onData(this._onTerminalSendString.bind(this));
```
can we replace it to `this.terminal.onData(() => this._onTerminalSendString.bind());` ?

KhizerRehan replied:
2 points here:
- `bind` takes `this` argument which is missing.
- we can use like this passing as a `arrow` function call back which is `lexically` using parent `this` no need to use `this` bind so I just used shorter syntax. 

```ts
    this.terminal.onData(event => {
      this._onTerminalSendString(event);
    });
```

Talha-Jamil-TJ commented on src/app/shared/components/xterm-terminal/terminal-status-bar/template.html:
```diff
@@ -0,0 +1,50 @@
+       fxLayoutAlign="end center"
+       fxLayoutGap="10px">
+    <button *ngIf="showCloseButton"
+            mat-icon-button
```
please add `id` to the button for unit testing

Waseem826 commented on src/app/shared/components/terminal/component.ts:
```diff
@@ -0,0 +1,293 @@
+  isLoadingTerminal = true;
+  isDexAuthenticationPageOpened = false;
+
+  showCloseButtonOnToolbar;
+  showOpenInSeparateViewButtonOnToolbar;
+  showCloseButtonOnStatusToolbar;
```
Missing types.

Waseem826 commented on src/app/shared/components/terminal/template.html:
```diff
@@ -0,0 +1,48 @@
+  </div>
+</ng-container>
+
+<ng-template #showTerminal>
```
```suggestion
<ng-template #terminalContainer>
```

Waseem826 commented on src/app/shared/components/terminal/component.ts:
```diff
@@ -0,0 +1,293 @@
+        this.message = 'Please wait, provisioning your Cloud Shell machine...';
+      } else if (frame.Data === ErrorOperations.connectionPoolExceeded) {
+        this.terminal.write(
+          `Oops! There could be ${this.MAX_SESSION_SUPPORTED} concurrent session per user either discard or close other sessions`
```
```suggestion
          `Oops! There could be ${this.MAX_SESSION_SUPPORTED} concurrent session per user. Please either discard or close other sessions.`
```

Waseem826 commented on src/app/shared/components/terminal/style.scss:
```diff
@@ -0,0 +1,48 @@
+// By wrapping styles with "selector" will prevent styles to go up the chain to parent and sibling components.
+// It will only effect styles specified here will go down to children components only.
+
+km-terminal {
```
Please use class instead of component selector.

Waseem826 commented on src/app/shared/components/terminal/template.html:
```diff
@@ -0,0 +1,48 @@
+                          (close)="onClose()"></km-terminal-status-bar>
+
+  <div #terminal
+       id="test"
```
Why `test` id?

Waseem826 commented on src/app/shared/components/terminal/template.html:
```diff
@@ -0,0 +1,48 @@
+
+  <div #terminal
+       id="test"
+       content></div>
```
Just curious what does `content` do?

Waseem826 commented on src/app/shared/components/terminal/terminal-status-bar/style.scss:
```diff
@@ -0,0 +1,45 @@
+@use 'mixins';
+
+.toolbar-status-container {
+  @include mixins.size(100%, 3rem);
+
+  align-items: center;
+  box-sizing: border-box;
+  display: flex;
+  padding: 0 1.75rem;
+
+  & button[mat-flat-button],
+  & button[mat-icon-button], {
+    @include mixins.size(auto, 1.8rem);
```
We don't use relative size units (`em`, `rem` etc.) so any particular reason for using them here?

Waseem826 commented on src/assets/css/_images.scss:
```diff
@@ -394,6 +398,38 @@
+  }
+}
+
+.km-icon-quota {
+  @include mixins.mask-image('/assets/images/icons/quota.svg', 24px);
+}
+
+.km-icon-helm {
+  @include mixins.provider-image('/assets/images/icons/helm.svg', 24px);
+}
+
+.km-icon-git {
+  @include mixins.provider-image('/assets/images/icons/git.svg', 24px);
+}
+
+.km-icon-namespace {
+  @include mixins.mask-image('/assets/images/icons/namespace.svg', 24px);
+}
+
+.km-icon-application {
+  @include mixins.mask-image('/assets/images/icons/application.svg', 24px);
+}
+
```
Duplicate styles

Waseem826 commented on src/assets/themes/dark.scss:
```diff
@@ -78,7 +88,7 @@ $foreground: map.merge($foreground, (disabled-text: #636363));
 $foreground: map.merge($foreground, (disabled-checkbox-checked: #7e868f));
 $foreground: map.merge($foreground, (disabled-checkbox-border: #b6b6b6));
-$foreground: map.merge($foreground, (disabled: #7c7c7c));
+$foreground: map.merge($foreground, (disabled: #313132));
```
Why do we need to change this? This was changed to `#7c7c7c` in https://github.com/kubermatic/dashboard/pull/5023

Waseem826 reviewed (changes requested):
Some feedback. PTAL

KhizerRehan commented:
/test pre-dashboard-test-unit

Waseem826 reviewed (approved):
/approve

Waseem826 commented:
/unhold
